src/fparser/two/parser.py

```diff
@@ -197,91 +180,103 @@ def _setup(self, input_classes):
 
         """
+        # pylint: disable=import-outside-toplevel
+        from fparser.two import Fortran2003
 
-        __autodoc__ = []
-        base_classes = {}
-
-        import logging
-        import fparser.two.Fortran2003
-
-        class_type = type(fparser.two.Fortran2003.Base)
+        class_type = type(Fortran2003.Base)
 
         # Reset subclasses dictionary in case this function has been
         # called before. If this is not done then multiple calls to
         # the ParserFactory create method may not work correctly.
-        fparser.two.Fortran2003.Base.subclasses = {}
+        Fortran2003.Base.subclasses = {}
+        base_classes = {}
 
-        for clsinfo in input_classes:
-            clsname = "{0}.{1}".format(clsinfo[1].__module__, clsinfo[0])
-            cls = eval(clsname)
+        for _, cls in input_classes:
             # ?? classtype is set to Base so why have issubclass?
             if (
                 isinstance(cls, class_type)
-                and issubclass(cls, fparser.two.Fortran2003.Base)
+                and issubclass(cls, Fortran2003.Base)
                 and not cls.__name__.endswith("Base")
             ):
                 base_classes[cls.__name__] = cls
-                if len(__autodoc__) < 10:
-                    __autodoc__.append(cls.__name__)
 
-        #
         # OPTIMIZE subclass_names tree.
         #
-
-        if 1:  # Optimize subclass tree:
-
-            def _rpl_list(clsname):
-                if clsname not in base_classes:
-                    error_string = "Not implemented: {0}".format(clsname)
-                    logging.getLogger(__name__).debug(error_string)
-                    return []
-                # remove this code when all classes are implemented.
-                cls = base_classes[clsname]
-                if hasattr(cls, "match"):
-                    return [clsname]
-                bits = []
-                for names in getattr(cls, "subclass_names", []):
-                    list1 = _rpl_list(names)
-                    for names1 in list1:
-                        if names1 not in bits:
-                            bits.append(names1)
-                return bits
-
-            for cls in list(base_classes.values()):
-                if not hasattr(cls, "subclass_names"):
-                    continue
-                opt_subclass_names = []
-                for names in cls.subclass_names:
-                    for names1 in _rpl_list(names):
-                        if names1 not in opt_subclass_names:
-                            opt_subclass_names.append(names1)
-                if not opt_subclass_names == cls.subclass_names:
-                    cls.subclass_names[:] = opt_subclass_names
-
-        # Initialize Base.subclasses dictionary:
-        for clsname, cls in list(base_classes.items()):
-            subclass_names = getattr(cls, "subclass_names", None)
-            if subclass_names is None:
+        def _rpl_list(clsname):
```

Comment:
Can we name it something more verbose/descriptive than "rpl"?

Reply:
I realised after I put this in that the reviewer would (quite rightly) complain about the fact I hadn't changed the name. It's now `_closest_descendants_with_match`.